### DISCUSSION TITLE
"Modify the following as needed"

### DISCUSSION THREAD

**Discussion by tpict** [fzyzcjy/flutter_rust_bridge#1369] [Q&A] [answered]
I'm following the setup guide, and [I'm stuck here](https://cjycode.com/flutter_rust_bridge/library/flutter_wrapper.html)

> Finally, we will need to write some code to be able to handle FFI in Flutter. Modify the following as needed (replacing library_name and LibraryName with your library name).

Do I need to replace the contents of those files set up by [the setup script](https://cjycode.com/flutter_rust_bridge/library/dart_only.html), or should I keep the `createWrapperImpl`, `createWrapper` definitions that are already there?

**Comment by fzyzcjy** [comment]
https://github.com/fzyzcjy/flutter_rust_bridge/blame/d54eae758f94ca1a5e69787d20f16af39daa89c2/book/src/library/flutter_wrapper.md#L37

cc @GregoryConrad who PRed that helpful part of doc - shall we clarify a bit in the doc?

**Comment by GregoryConrad** [comment]
> Do I need to replace the contents of those files set up by [the setup script](https://cjycode.com/flutter_rust_bridge/library/dart_only.html),

@tpict You were intended to modify the two variable declarations at the top of the setup script so you wouldn't need to manually modify any of those files after.

> or should I keep the createWrapperImpl, createWrapper definitions that are already there?

Keep those definitions, but change the `LibraryName` and `library_name` parts in the definitions to your library name.

**Reply by tpict** [reply]
appreciate it! I did modify the variables in the setup script. My question was whether or not I should keep the contents of those files when setting up the Flutter wrapper, or just replace them with what's written out in the guide (swapping `LibraryName` etc as needed)

**Comment by tpict** [comment]
I _think_ the answer is to keep the existing content, though it seems that the generated classes aren't respecting the `DART_CLASS_NAME` variable in the setup script, e.g. if I have
```
LIBNAME=helloworld # snake_case
DART_CLASS_NAME=HelloWorld # probably is PascalCase version of $LIBNAME
```

then `ffi.dart` etc references `HelloWorld`, which I have to update to `Helloworld` to successfully build.

When I actually call `createLib` on my Mac, I get `Failed to lookup symbol 'store_dart_post_cobject'`. [I found the relevant section in the guide](https://cjycode.com/flutter_rust_bridge/troubleshooting.html#issue-with-store_dart_post_cobject) but it hasn't resolved the issue for me. What I'm trying to do is: follow the "creating a Dart/Flutter library" section to make a Flutter package I can install in another project, but I want to install it from that package locally using the `path:` directive in the project's pubspec.yaml so I can test it out. I've copied the .xcframework.zip to the macos dir as directed.

edit again: I also tried moving my project in a Melos monorepo with my Dart/Flutter libraries instead of using the `path`: directive and I see the same `Failed to lookup symbol 'store_dart_post_cobject'` issue. Searching the repo I see a lot of solutions around tweaking `Runner.xcodeproj`, but that's not present if you follow the Dart/Flutter library guide

**Reply by GregoryConrad** [reply]
> `Failed to lookup symbol 'store_dart_post_cobject'`

This means the binary isn't being included in the build. The zip file should be in `macos/Frameworks` and/or `ios/Frameworks`, with a very specific name: `library_name-v0.0.0.zip` (replacing 0.0.0 with the version found in the pubspec).

For reference, the name you are looking for can be found in the first line of the `.podspec` file (should be `release_tag_name = ...`. Your zip name should be that + `.zip`.

Unfortunately, this process is really error-prone and I hate it. So much can go wrong, and too easily. Unfortunately, we can't do too much better until Native Assets come around.

**Reply by tpict** [reply]
Yep, that's exactly what I have. `macos/Frameworks/hello_world-v1.1.1.zip` is present and the podspec references 
```
release_tag_name = 'hello_world-v1.1.1' # generated; do not edit
```

I did start over to make sure `LIBNAME` and `DART_CLASS_NAME` were _strictly_ snake-case/PascalCase versions of each other which fixed the Dart base, but still no joy with `Failed to lookup symbol 'store_dart_post_cobject'`

**Reply by GregoryConrad** [reply]
Have you run melos bootstrap?

**Reply by tpict** [reply]
Yeah. I do something like
```
cargo build
melos build:apple
cp platform-build/HelloWorld.xcframework.zip packages/flutter_hello_world/macos/Frameworks/hello_world-v1.1.1.zip
melos bs
```

**Reply by GregoryConrad** [reply]
@tpict Do you have a copy of this on GitHub or somewhere else I could take a look at the packages

**Reply by tpict** [reply]
I’ll set up a minimal reproduction for you ASAP

**Reply by tpict** [reply]
@GregoryConrad Here it is: https://github.com/tpict/flutter_rust_bridge_troubleshooting/tree/main

Running above steps (version is 1.1.0 for this minimal reproduction) and then running `flutter test` in `my_app` results in the dreaded `symbol not found`

**Reply by GregoryConrad** [reply]
@tpict The example repo you gave is working fine for me on macOS after:
1. Adding `hello_world` as a dependency in `my_app` (not sure if this is actually needed)
2. Changing `spec.name = 'hello_world'` to `spec.name  = 'flutter_hello_world'` in both iOS/macOS podspecs

**Reply by tpict** [reply]
@GregoryConrad Still no joy after making those changes. Are there any useful diagnostics I can provide?

**Reply by tpict** [reply]
Still trying to work around this. Are there any "absolute beginners" docs on how `DynamicLibrary.executable()` works? I can see where `store_dart_post_cobject` is declared–I just don't understand why `dart:ffi` isn't finding it.

**Reply by GregoryConrad** [reply]
Sorry, this escaped me.

Iirc by default the library docs use static linking on iOS/macOS and dynamic loading elsewhere. Thus, the slightly misnamed DynamicLibrary.executable() actually returns the exec file itself, which should have the requested ffi objects if static linking with the rust library was performed.

I’m 99% sure it’s just because it’s not getting linked in. Not sure why, but that’s likely it. Maybe tree shaking or something funky like that?

If this isn’t time sensitive, it might be worth waiting for native assets. If  you need this on the sooner end though, then native assets won’t help.

**Reply by tpict** [reply]
It sounds like native assets would make things easier. But it would be nice to be able to get started on the Rust side in the meantime 🙂 

I checked `libhello_world.a` and it does contain the expected symbol:
```
0000000000000280 T _store_dart_post_cobject
```

so I think it's been statically linked correctly? (I'm not a C wiz)

**Reply by GregoryConrad** [reply, marked as answer]
You don’t want to check the .a; that’ll have the symbols if the rust compiler does its job.

what you’d want to check is the executable of the application itself (the one hidden with the .app bundle). However, I can tell you already that it doesn’t have the symbols (unless the Dart language ffi implementation is wrong, but I highly doubt that). For some reason, the .a is not getting statically linked into the executable. As for why, that’s a bit harder to answer. It’s either not getting found from the podspec or it’s getting tree shaken.

Could you try to see if the example app under flutter_mimir/example here works for you? https://github.com/GregoryConrad/mimir

Melos bs, Melos run build:apple, copy the framework zip over to frameworks folder with the right name, and then finally flutter run in that flutter mimir example 

(Sorry for terrible formatting—on phone)

**Reply by tpict** [reply]
I get a lot of complaints of
```
ld: warning: object file (/Users/tpict/Projects/mimir/packages/flutter_mimir/example/build/macos/B
uild/Products/Debug/XCFrameworkIntermediates/flutter_mimir/libembedded_milli.a(debug.o)) was built
 for newer macOS version (12.0) than being linked (10.14)
```

but the app works!

...and it finally clicked: I missed the section on the site about unit testing. I just moved the test in my `hello_world` app to be an integration test and now it passes. My bad. Thanks so much for all your help 🙏

**Reply by GregoryConrad** [reply]
Ahh alright! If the integration test on macOS/iOS passes, then it definitely should be statically linking correctly, since integration tests use the full build process. Glad to hear it is working!
